autopilot/externalscoreattach.go

```diff
@@ -0,0 +1,124 @@
+package autopilot
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/btcsuite/btcutil"
+)
+
+// ScoreSettable is an interface that indicates that the scores returned by the
+// heuristic can be set by an external caller. The ExternalScoreAttachment
+// imlements this interface, and so should any heuristic that is using the
+// ExternalScoreAttachment as a sub-heuristic, to allow access to setting the
+// internal scores.
+type ScoreSettable interface {
```

Comment:
non-blocking naming suggestion: `MutableHeuristic`?

TBH, my preferred alternative is actually to roll this into the `AttachmentHeuristic` interface, and non-mutable heuristics just return nil. This avoids the type casting and the targeting issues discussed elsewhere

Reply:
If we keep the additional interface, it belongs in interface.go

Reply:
Moving it to the `AttachmentHeuristic` interface would cause all implementation to have to implement this new method even if they don't actually utilize it. Splitting up the interface creates a specific interface which can be targeted precisely via type assertions. I don't think it's the case that type assertions are inherently "bad", they're simply a feature of the language to upgrade/downgrade interfaces.

Reply:
> I don't think it's the case that type assertions are inherently "bad", they're simply a feature of the language to upgrade/downgrade interfaces.

Certainly, type assertions have their place. IME however an assertion from interface->interface typically indicates that the abstractions are insufficient or surrounding logic is overloaded, even if using a type assertion readily enables compilation. 

Side note: not that performance is a huge concern in this particular instance, but interface to interface conversion is also the most expensive variant by a significant margin.

> Moving it to the `AttachmentHeuristic` interface would cause all implementation to have to implement this new method even if they don't actually utilize it.

How is that different that forcing `WeightedCombAttachment` to implement the `ScoreSettable` interface, even though it is it not the source of the scores it outputs? Its output is purely a dervied value. Unless calling `SetNodeScores` caused it to override the weighted combination (which is what is required as per the godoc comment on `SetNodeScores`), idt it makes sense to force it into the `ScoreSettable` interface. 

This also breaks the abstraction at the interface level, since you can't determine the behavior of calling `SetNodeScores` on a `ScoreSettable` without also knowing or asserting which `AttachmentHeuristic` you're calling it on: i.e. will it return exactly the scores I tell it, or will it return a dervied/weighted score for which these are only a component?

Perhaps a solution is to differentiate between a source (either static or mutable, conceptually a map) and a heuristic (derived from one or more sources). For instance:
```golang
type MutableHeuristicSource interface {
        AttachmentHeuristic

        SetNodeScores(map[NodeID]float64) error
}
...
var _ MutableHeuristicSource = (*ExternalScoreAttachment)(nil)
```

Since `ExternalScoreAttachment` then implements `AttachmentHeuristic` as a subinterface it can be used directly to instantiate `NewWeightedCombAttachment` as is the case right now. However, the combined interface lets us retain the same pointer with access to mutate the underlying source.

In addition to accepting the active `Heuristic` within the manager's config, we also accept an optional `MutableSources []MutableHeuristicSource` parameter. Upon initialization, the manager constructs an index of `map[source.Name()]  = source`. The method for updating scores then looks like:
```golang
func (m *Manager) SetSourceScores(name string, scores map[NodeID]float64) error {
	source, ok := m.mutableSourceIndex[name]
	if !ok {
		return ErrUnknownSource
	}

 	return source.SetNodeScores(scores)
}
```
where `name` can be additional parameter passed via rpc to target a specific source. Since the same object is referenced inside the `WeightedCombAttachment` we can safely update its scores without any additional coordination besides the mutex internal to `ExternalScoreAttachment`.

The major benefits here are that derived heuristics won't need to implement `SetNodeScores` unless they are 1) pure source or 2) we actually want to be able to override them. In addition, `WeightedCombAttachment` does not need to be modified at all and better isolates the heuristic derivation pipeline from the actual values flowing through it. Lastly, it provides a way of target specific mutable heuristics within the pipline without having to squeeze that logic through higher-level heuristics via `SetNodeScore`, which otherwise doesn't take any sort of identifier.

Reply:
Good suggestions :) I do see the benefit of being able to target a specific sub-heuristic to set the scores. 

Note that the API currently assumes (at least the few docs that are there) that the scores you set are also the scores the heuristic returns. Good point that in the case of the `Weighted` heuristic this won't be the exact case.

By specifying a name, it could be more clear that you are actually setting the scores on the `externalScore` sub-heuristic, and trying to target the `WeightedComb` directly would fail.

In regards to how to structure the interfaces etc, I don't have any strong opinions about it, I just used what was easiest to get this added. Since this is an experimental sub-server RPC, I think we can just start with something that works for our use case, and then modify it recklessly if we figure out there are better ways to organize it once we have more `ScoreSettable` sub-heuristics (if that even makes sense).

Reply:
Defined a new `SuperHeuristic` that can be used to get all sub-heuristics. From there we now target the name of the sub-heuristic as suggested.